### ISSUE TITLE
Types not taken into account when coming from "types" field in libraries' package.json

### ISSUE THREAD

**Issue by bryanvanwijk** [palantir/tslint#3931]
### Bug Report

- __TSLint version__: 5.9.1
- __TypeScript version__: 2.5.3
- __Running TSLint via__: CLI (`tslint -c tslint.json -p .`)

We have an i18n project which provides common functions to other projects, but TSlint does not warn about unhandled promises from these functions when we use them in other projects. The types are exposed using `"types": "build/src/index.d.ts"` in the package.json of the i18n project. TSLint does warn about unhandled promises if we expose the functions using an `index.d.ts` file in the root directory without the `types` property in the package.json file.

#### TypeScript code being linted

```ts
import { downloadAllLanguages,  } from "i18n";

downloadAllLanguages(["en", "de", "nl"], "./data/i18n"); // Returns a promise but tslint does not warn about this.
```

with `tslint.json` configuration:

```json
{
  "extends": "tslint:latest",
  "rules": {
    "object-literal-key-quotes": [true, "as-needed"],
    "no-trailing-whitespace": false,
    "no-angle-bracket-type-assertion": false,
    "no-conditional-assignment": false,
    "radix": false,
    "no-empty-interface": false,
    "prefer-const": false,
    "curly": true,
    "ordered-imports": [
      false
    ],
    "max-line-length": [
      true,
      250
    ],
    "max-classes-per-file": [
      true,
      3
    ],
    "variable-name": [
      false
    ],
    "trailing-comma": [
      false
    ],
    "arrow-parens": false,
    "comment-format": [
      false
    ],
    "object-literal-sort-keys": false,
    "eofline": false,
    "whitespace": [
      true,
      "check-branch",
      "check-decl",
      "check-operator",
      "check-separator",
      "check-type"
    ],
    "no-unused-variable": [
      true
    ],
    "no-submodule-imports": false,
    "member-ordering": false,
    "prefer-conditional-expression": false,
    "await-promise": [true, "Thenable", "Bluebird", "Promise", "RequestPromise"],
    "semicolon": [true, "ignore-bound-class-methods"],
    "no-floating-promises": [true, "Thenable", "Bluebird", "Promise", "RequestPromise"],
    "interface-over-type-literal": false,
    "interface-name": false,
    "no-implicit-dependencies": [true, "dev"],
    "no-string-throw": true
  },
  "jsRules": {
    "curly": true
  },
  "rulesDirectory": [
  ]
}
```

#### Actual behavior
TSLint does not warn about the unhandled promises from imported functions which expose their typings using the types property in `package.json`

#### Expected behavior
TSLint should raise a `Promises must be handled appropriately` error for functions from modules which expose their typings with `"types": "build/src/index.d.ts"` in the package.json file.

**Comment by JKillian** [comment]
Hmm, sounds to me less an issue with the `no-floating-promises` rule and more an issue with how TSLint sets things up for type-checking. I renamed the issue to reflect this. Agree that it should work in both cases you describe above though!

**Comment by JoshuaKGoldberg** [comment]
☠️ *TSLint's time has come!* ☠️ 

TSLint is no longer accepting most feature requests per #4534. See **[typescript-eslint.io](https://typescript-eslint.io)** for the new, shiny way to lint your TypeScript code with **ESLint**. ✨ 

It was a pleasure open sourcing with you all!

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈 _(#4534)_  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
